cpp/src/copying/contiguous_split.cu

```diff
@@ -14,466 +14,1129 @@
- * of the offset columns for all input string columns and computes this information with a single
- * thrust call.  In addition, the vector returned is allocated for -all- columns in the table so
- * further processing of non-string columns can happen afterwards.
- *
- * The key things this function avoids
- * - avoiding reaching into gpu memory on the cpu to retrieve offsets to compute string sizes.
- * - creating column_device_views on the base string_column_view itself as that causes gpu memory
- * allocation.
+ * @brief Functor that computes the number of buffers for a single column.
+ *
+ * Called by count_src_bufs to compute buffer count for single source column.  This function will
+ * recursively call count_src_bufs in the case of nested types.
  */
-thrust::host_vector<column_split_info> preprocess_string_column_info(
-  cudf::table_view const& t,
-  rmm::device_vector<column_split_info>& device_split_info,
-  cudaStream_t stream)
+struct buf_count_functor {
+  // strings column : offsets, chars, potentially validity
+  template <typename T, std::enable_if_t<std::is_same<T, cudf::string_view>::value>* = nullptr>
+  size_type operator()(column_view const& col)
+  {
+    return 2 + (col.nullable() ? 1 : 0);
+  }
+
+  // nested types : # of children + potentially validity
+  template <typename T,
+            std::enable_if_t<std::is_same<T, cudf::list_view>::value or
+                             std::is_same<T, cudf::struct_view>::value>* = nullptr>
+  size_type operator()(column_view const& col)
+  {
+    return count_src_bufs(col.child_begin(), col.child_end()) + (col.nullable() ? 1 : 0);
+  }
+
+  template <typename T, std::enable_if_t<std::is_same<T, cudf::dictionary32>::value>* = nullptr>
+  size_type operator()(column_view const& col)
+  {
+    CUDF_FAIL("Unsupported type");
+  }
+
+  // plain types : 1 + potentially validity
+  template <typename T, std::enable_if_t<!cudf::is_compound<T>()>* = nullptr>
+  size_type operator()(column_view const& col)
+  {
+    return 1 + (col.nullable() ? 1 : 0);
+  }
+};
+
+template <typename InputIter>
+size_type count_src_bufs(InputIter begin, InputIter end)
 {
-  // build a list of all the offset columns and their indices for all input string columns and put
-  // them on the gpu
-  thrust::host_vector<column_preprocess_info> offset_columns;
-  offset_columns.reserve(t.num_columns());  // worst case
-
-  // collect only string columns
-  size_type column_index = 0;
-  std::for_each(t.begin(), t.end(), [&offset_columns, &column_index](cudf::column_view const& c) {
-    if (c.type().id() == type_id::STRING) {
-      cudf::column_device_view cdv((strings_column_view(c)).offsets(), 0, 0);
-      offset_columns.push_back(
-        column_preprocess_info{column_index, c.offset(), c.size(), c.has_nulls(), cdv});
-    }
-    column_index++;
+  auto buf_iter = thrust::make_transform_iterator(begin, [](column_view const& col) {
+    return cudf::type_dispatcher(col.type(), buf_count_functor{}, col);
```

Comment:
Shouldn't this functionality be type agnostic? It can be inferred by just traversing the column's children and checking nullability of each.

Reply:
This doesn't work because not every type actually contains data itself (fixed width column does but strings column doesn't).  I suppose I could check for the data buffer of the column itself being empty but that feels a little weak.

Reply:
Checking if the data pointer is null doesn't seem strange at all. It feels strange to type specialize functionality that is essentially just traversing a tree and counting the number of non-null pointers. This is clearer, more compact, and insulated against adding future column types (like dictionary).